File: ISS Emergency/Assets/Scripts/LayerSetter.js
#pragma strict


// Hannes Helmholz
//
// 


@script RequireComponent(NetworkView) // for RPC
@script RequireComponent(UniqueIDManager)

private class LayerSetter extends MonoBehaviour {
	
	// =============================================================================
	// MEMBERS ---------------------------------------------------------------------
	private static var NetView:NetworkView;
	
	private var IDManager:UniqueIDManager;
	// =============================================================================
	
	
	
	// =============================================================================
	// METHODS UNITY ---------------------------------------------------------------
	
	function Awake() {
		NetView = networkView;
		
		IDManager = gameObject.GetComponent.<UniqueIDManager>();
	}
	// =============================================================================
	
	
	
	// =============================================================================
	// METHODS STATIC --------------------------------------------------------------
	
	public static function SetAllLayerRecursively(go:GameObject, newLayer:int) {
		SetLayer(go, newLayer);
		
		NetView.RPC("RPCSetAllLayerRecursively", RPCMode.OthersBuffered, go.GetComponent.<UniqueID>().ID, newLayer);
	}
	
	
	private static function SetLayer(go:GameObject, newLayer:int) {
	    if (!go)
	    	return;
	
	    go.layer = newLayer;
	    
	    for (var child:Transform in go.transform) {
	        if (!child)
	        	continue;
	        	
	        SetLayer(child.gameObject, newLayer);
	    }
	}
	// =============================================================================
	
	
	
	// =============================================================================
	// METHODS RPC -----------------------------------------------------------------
	
	@RPC
	function RPCSetAllLayerRecursively(id:int, newLayer:int) {
		SetLayer(IDManager.GetGameObjectByID(id), newLayer);
	}
	// =============================================================================
}

File: Assets/Scripts/LayerSetter.cs

// Hannes Helmholz
//
// 

using UnityEngine;

[RequireComponent (typeof(NetworkView))] // for RPC
[RequireComponent (typeof(UniqueIDManager))]


//organisiert game-objekten und zugehoerige ebene
public class LayerSetter : MonoBehaviour
{
 
    // =============================================================================
    // MEMBERS ---------------------------------------------------------------------
    private static NetworkView NetView;
    private UniqueIDManager IDManager;
    // =============================================================================
 
 
 
    // =============================================================================
    // METHODS UNITY ---------------------------------------------------------------
 
    void Awake ()
    {
        NetView = networkView;
     
        IDManager = gameObject.GetComponent<UniqueIDManager> ();
    }
    // =============================================================================
 
 
 
    // =============================================================================
    // METHODS STATIC --------------------------------------------------------------
 
    public static void SetAllLayerRecursively ( GameObject go, int newLayer )
    {
        SetLayer ( go, newLayer );
     
        NetView.RPC ( "RPCSetAllLayerRecursively", RPCMode.OthersBuffered, go.GetComponent<UniqueID> ().ID, newLayer );
    }
 
 
    private static void SetLayer ( GameObject go, int newLayer )
    {
        if ( !go )
        {
            return;
        }
 
        go.layer = newLayer;
     
        foreach ( Transform child in go.transform )
        {
            if ( !child )
            {
                continue;
            }
             
            SetLayer ( child.gameObject, newLayer );
        }
    }
    // =============================================================================
 
 
 
    // =============================================================================
    // METHODS RPC -----------------------------------------------------------------
 
    [RPC]
    void RPCSetAllLayerRecursively ( int id, int newLayer )
    {
        SetLayer ( IDManager.GetGameObjectByID ( id ), newLayer );
    }
    // =============================================================================
}